src/Confluent.Kafka/Impl/SafeTopicHandle.cs

```diff
@@ -88,28 +88,83 @@ internal long Produce(byte[] val, int valOffset, int valLength, byte[] key, int
-                        (IntPtr) (MsgFlags.MSG_F_COPY | (blockIfQueueFull ? MsgFlags.MSG_F_BLOCK : 0)),
-                        pValue, (UIntPtr) valLength,
-                        pKey, (UIntPtr) keyLength,
-                        opaque);
+                    gchValue.Free();
                 }
-                else
+
+                if (key != null)
+                {
+                    gchKey.Free();
+                }
+            }
+        }
+
+        //producev return an errorCode where produce return 0 or -1 with error retrievable in LastError
+        //duplicates marshal oprations as this part is critical
+        //TODO refactor if benchmark show it's ok
+        //TODO also benchmark produce vs producev, with an override of producev without timestamp
+        internal ErrorCode Producev(byte[] val, int valOffset, int valLength, byte[] key, int keyOffset, int keyLength, int partition, long timestamp, IntPtr opaque, bool blockIfQueueFull)
+        {
+            var pValue = IntPtr.Zero;
+            var pKey = IntPtr.Zero;
+
+            var gchValue = default(GCHandle);
+            var gchKey = default(GCHandle);
+
+            if (val == null)
+            {
+                if (valOffset != 0 || valLength != 0)
                 {
-                    return (long) LibRdKafka.producev(
-                        handle,
-                        partition,
-                        (IntPtr) (MsgFlags.MSG_F_COPY | (blockIfQueueFull ? MsgFlags.MSG_F_BLOCK : 0)),
-                        pValue, (UIntPtr) valLength,
-                        pKey, (UIntPtr) keyLength,
-                        timestamp.Value,
-                        opaque);
+                    throw new ArgumentException("valOffset and valLength parameters must be 0 when producing null values.");
                 }
             }
+            else
+            {
+                gchValue = GCHandle.Alloc(val, GCHandleType.Pinned);
+                pValue = Marshal.UnsafeAddrOfPinnedArrayElement(val, valOffset);
+            }
+
+            if (key == null)
+            {
+                if (keyOffset != 0 || keyLength != 0)
+                {
+                    throw new ArgumentException("keyOffset and keyLength parameters must be 0 when producing null key values.");
+                }
+            }
+            else
+            {
+                gchKey = GCHandle.Alloc(key, GCHandleType.Pinned);
+                pKey = Marshal.UnsafeAddrOfPinnedArrayElement(key, keyOffset);
+            }
+
+            try
+            {
+                // TODO: when refactor complete, reassess the below note.
+                // Note: since the message queue threshold limit also includes delivery reports, it is important that another
+                // thread of the application calls poll() for a blocking produce() to ever unblock.
+
+                return LibRdKafka.producev(
+                    kafkaHandle.DangerousGetHandle(),
```

Comment:
I don't like it but that was the simplest to implement first
@edenhill  should we put it directly in the SafeKafkaHandle and call it with string rk/topic rather than rk/rkt?